### PR TITLE
docs: document S3-compatible object storage for distributed checkpoint

### PR DESCRIPTION
**Pull request by goanpeca**
# Documentation and Public API Export

This change documents how to use Distributed Checkpoint (DCP) with S3 and S3-compatible object stores through the existing fsspec/s3fs storage layer. It also re-exports `FsspecReader` and `FsspecWriter` from `torch.distributed.checkpoint` so the new examples can use a public import path instead of the private `_fsspec_filesystem` module.

Before submitting, I reviewed:
- The Ultimate Guide to PyTorch Contributions
- The [AI-Assisted Development](https://github.com/pytorch/pytorch/blob/main/CONTRIBUTING.md#ai-assisted-development) policy

## What changed?

Documents how to point DCP at an S3-compatible object store using the existing fsspec/s3fs storage layer, so `FsspecReader`/`FsspecWriter` are usable with `s3://` URLs against Amazon S3 or any S3-compatible endpoint (Backblaze B2, Cloudflare R2, and MinIO are named as example providers) without changing DCP save/load behavior.

The PR also exposes `FsspecReader` and `FsspecWriter` from `torch.distributed.checkpoint`, updates the fsspec test import to cover the public path, and adds autodoc entries for the public classes.

## Details

`FsspecReader` and `FsspecWriter` forward extra keyword arguments to `fsspec.core.url_to_fs`, which for an `s3://` URL reach the [s3fs](https://s3fs.readthedocs.io/) filesystem. The docs did not spell this out, so the object-store path was easy to miss. The new section covers three things a user needs: installing the backend (`pip install s3fs`), supplying credentials (the standard `AWS_ACCESS_KEY_ID` / `AWS_SECRET_ACCESS_KEY` environment variables, or the s3fs `key`/`secret` arguments), and setting a custom endpoint for a non-AWS store via `client_kwargs={"endpoint_url": ...}` (Amazon S3 needs no endpoint). It shows matching save and load snippets with `FsspecWriter` and `FsspecReader`.

The fsspec-backed implementation now lazy-imports `fsspec` when resolving the URL, rather than at `torch.distributed.checkpoint` import time. This keeps `fsspec` optional for DCP users that do not use fsspec-backed storage while still letting Sphinx autodoc import the real `FsspecReader` and `FsspecWriter` classes and render their signatures/docstrings.

The wording is intentionally provider-neutral: it leads with Amazon S3, presents the endpoint as the general mechanism for any S3-compatible store, and names Backblaze B2, Cloudflare R2, and MinIO together as equal example providers. Endpoints in the snippets are placeholders (`https://your-s3-endpoint.example.com`). No provider is favored and no marketing content is included.

## Testing

- `cd docs && make html` builds cleanly and the new section renders in the Distributed Checkpoint page (the PR's docs-preview link can be used to verify the rendered output).
- `python -m py_compile torch/distributed/checkpoint/__init__.py torch/distributed/checkpoint/_fsspec_filesystem.py test/distributed/checkpoint/test_fsspec.py`
- `git diff --check -- torch/distributed/checkpoint/__init__.py torch/distributed/checkpoint/_fsspec_filesystem.py`
- `lintrunner -a` could not be run locally because `lintrunner` is not installed and no `.venv` exists in the repo root or parent directory.
- A runtime import check could not be run in this source checkout because the unbuilt tree is missing `torch.version`.
- The snippets follow the same illustrative style as the rest of this page (fenced code blocks alongside the autodoc directives) and use only public DCP entry points (`dcp.save`, `dcp.load`) with the documented `FsspecReader`/`FsspecWriter` keyword arguments.

## BC-breaking?

No. The public import is additive, `fsspec` remains optional, and no existing public API, behavior, or default is modified.

---

The commits are signed off under the DCO (`Signed-off-by:` trailers). This PR was prepared with the assistance of an AI coding tool; I have reviewed the content and am responsible for its accuracy.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pradeepfn @teja-rao @svekars @sekyondaMeta @AlannaBurke

### CODE REVIEW

File: docs/source/distributed.checkpoint.md [hunk 1]
```diff
@@ -234,6 +234,64 @@ We provide a filesystem based storage layer:
...
+    secret=os.environ["AWS_SECRET_ACCESS_KEY"],
+)
+dcp.save(state_dict=app_state, storage_writer=writer)
+```
+
+`FsspecReader` takes the same arguments for distributed loads:
+
+```python
+from torch.distributed.checkpoint._fsspec_filesystem import FsspecReader
+
```
[the remark this answers is not in the document]
Comment (goanpeca) [reply docs/source/distributed.checkpoint.md]: Fixed: Added the missing dcp import to the FsspecReader load example.

File: docs/source/distributed.checkpoint.md [hunk 2]
```diff
@@ -234,6 +234,64 @@ We provide a filesystem based storage layer:
...
+)
+dcp.save(state_dict=app_state, storage_writer=writer)
+```
+
+Credentials can also be passed explicitly instead of through the environment,
+using the s3fs `key` and `secret` arguments:
+
+```python
+import os
+
```
[the remark this answers is not in the document]
Comment (goanpeca) [reply docs/source/distributed.checkpoint.md:285]: Fixed: Added the missing dcp and FsspecWriter imports so the explicit credentials snippet is self-contained.

File: docs/source/distributed.checkpoint.md [hunk 3]
```diff
@@ -234,6 +234,64 @@ We provide a filesystem based storage layer:
...
+export AWS_SECRET_ACCESS_KEY="..."
+```
+
+For a non-AWS S3-compatible store, also pass its endpoint through the s3fs
+`client_kwargs` argument (AWS S3 needs no endpoint):
+
+```python
+import torch.distributed.checkpoint as dcp
+from torch.distributed.checkpoint._fsspec_filesystem import FsspecWriter
+
```
[the remark this answers is not in the document]
Comment (goanpeca) [reply docs/source/distributed.checkpoint.md:270]: Fixed: Re-exported FsspecReader and FsspecWriter from torch.distributed.checkpoint and updated the docs to use the public import.

File: torch/distributed/checkpoint/__init__.py [hunk 4]
```diff
@@ -2,6 +2,7 @@
 from .api import CheckpointException
 from .default_planner import DefaultLoadPlanner, DefaultSavePlanner
 from .filesystem import FileSystemReader, FileSystemWriter
+from ._fsspec_filesystem import FsspecReader, FsspecWriter
```
[the remark this answers is not in the document]
Comment (goanpeca) [reply torch/distributed/checkpoint/__init__.py]: Deferred: Updated the PR description to reflect the API and test changes; label edits require maintainer permissions.

File: docs/source/distributed.checkpoint.md [hunk 5]
```diff
@@ -234,6 +234,67 @@ We provide a filesystem based storage layer:
   :members:
 ```
 
+For object stores and other non-POSIX backends, DCP also ships an
+[fsspec](https://filesystem-spec.readthedocs.io/) backed storage layer.
+`FsspecReader` and `FsspecWriter` accept any URL that fsspec can resolve
+(for example `s3://bucket/path`) and forward any extra keyword arguments to
+`fsspec.core.url_to_fs`. For an `s3://` URL those arguments reach the
```
[the remark this answers is not in the document]
Comment (goanpeca) [reply docs/source/distributed.checkpoint.md:241]: Fixed: Added autoclass blocks for torch.distributed.checkpoint.FsspecReader and FsspecWriter.

File: docs/source/distributed.checkpoint.md [hunk 6]
```diff
@@ -234,6 +234,77 @@ We provide a filesystem based storage layer:
...
+
+```{eval-rst}
+.. autoclass:: torch.distributed.checkpoint.FsspecReader
+  :members:
+```
+
+```{eval-rst}
+.. autoclass:: torch.distributed.checkpoint.FsspecWriter
+  :members:
+```
```
[the remark this answers is not in the document]
Comment (goanpeca) [reply docs/source/distributed.checkpoint.md:255]: Fixed: Corrected the FsspecWriter docstring typo so the rendered autodoc text is accurate.

File: torch/distributed/checkpoint/__init__.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (goanpeca) [reply torch/distributed/checkpoint/__init__.py]: Fixed: Guarded the fsspec re-export so checkpoint imports without fsspec and raises a clear error only when constructing the fsspec APIs.

File: docs/source/distributed.checkpoint.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (goanpeca) [reply docs/source/distributed.checkpoint.md:255]: Fixed: Moved the fsspec import into lazy URL resolution so autodoc can import the real public classes without fsspec installed.

File: torch/distributed/checkpoint/__init__.py [hunk 7]
```diff
@@ -19,3 +19,23 @@
...
+    if exc.name != "fsspec":
+        raise
+
+    _fsspec_import_error = exc
+
+    class FsspecReader:
+        def __init__(self, *args: object, **kwargs: object) -> None:
+            raise ModuleNotFoundError(
+                "FsspecReader requires fsspec. Install it with `pip install fsspec`."
+            ) from _fsspec_import_error
```
[the remark this answers is not in the document]
Comment (goanpeca) [reply torch/distributed/checkpoint/__init__.py]: Fixed: Removed the fallback FsspecReader stub, so autodoc now sees the real class and docstring.
